File: 1d-spreadsheet.cpp
#include <iostream>
#include <string>
#include <vector>

struct cell
{
  std::string op;
  std::string lhs;
  std::string rhs;
};
  
int main()
{
  int n;
  std::cin >> n;

  std::vector<cell> cells(n);
  
  for (cell& c : cells)
    std::cin >> c.op >> c.lhs >> c.rhs;

  std::vector<int> queue;
  queue.reserve(n);

  std::vector<int> eval_queue;
  eval_queue.reserve(n);
  
  std::vector<bool> in_queue(n, false);
  
  for (int i(0); i!=n; ++i)
    {
      if (in_queue[i])
        continue;

      queue.emplace_back(i);
      in_queue[i] = true;

      while (!queue.empty())
        {
          const int index(queue.back());
          const cell& cell(cells[index]);
          bool ready(true);

          if (cell.lhs[0] == '$')
            {
              const int ref(std::stoi(cell.lhs.substr(1)));

              if (!in_queue[ref])
                {
                  ready = false;
                  queue.emplace_back(ref);
                  in_queue[ref] = true;
                }
            }

          if (cell.rhs[0] == '$')
            {
              const int ref(std::stoi(cell.rhs.substr(1)));

              if (!in_queue[ref])
                {
                  ready = false;
                  queue.emplace_back(ref);
                  in_queue[ref] = true;
                }
            }

          if (ready)
            {
              queue.pop_back();
              eval_queue.emplace_back(index);
            }
        }
    }

  std::vector<long> result(n);
  
  auto value_of =
    [&](const std::string& arg) -> long
    {
      if (arg[0] == '$')
        {
          const int ref(std::stoi(arg.substr(1)));
          return result[ref];
        }
      else
        return std::stol(arg);
    };

  for (int i : eval_queue)
    {
      const cell& cell(cells[i]);

      if (cell.op == "VALUE")
        result[i] = value_of(cell.lhs);
      else if (cell.op == "ADD")
        result[i] = value_of(cell.lhs) + value_of(cell.rhs);
      else if (cell.op == "SUB")
        result[i] = value_of(cell.lhs) - value_of(cell.rhs);
      else if (cell.op == "MULT")
        result[i] = value_of(cell.lhs) * value_of(cell.rhs);
    }

  for (long r : result)
    std::cout << r << '\n';

  return 0;
}

File: 1d-spreadsheet.py
import math
import sys

n = int(input())
in_eval_queue = list(False for i in range(n))
result = list(0 for i in range(n))
eval_queue = []
ops = []

for i in range(n):
    operation, lhs, rhs = input().split()
    ops.append((operation, lhs, rhs))

for i in range(n):
    if in_eval_queue[i]:
        continue
    
    queue = [i]
    
    while queue:
        cell = queue[-1]

        if in_eval_queue[cell]:
            queue.pop()
            continue
        
        op, lhs, rhs = ops[cell]
        ready = True
    
        if lhs[0] == '$':
            ref = int(lhs[1:])
            if not in_eval_queue[ref]:
                ready = False
                queue.append(ref)
            
        if rhs[0] == '$':
            ref = int(rhs[1:])
            if not in_eval_queue[ref]:
                ready = False
                queue.append(ref)

        if ready:
            queue.pop()
            eval_queue.append(cell)
            in_eval_queue[cell] = True

def value_of(arg):
    if arg[0] == '$':
        return result[int(arg[1:])]
    else:
        return int(arg)
    
for i in eval_queue:
    op, lhs, rhs = ops[i]

    if op == "VALUE":
        result[i] = value_of(lhs)
    elif op == "ADD":
        result[i] = value_of(lhs) + value_of(rhs)
    elif op == "SUB":
        result[i] = value_of(lhs) - value_of(rhs)
    elif op == "MULT":
        result[i] = value_of(lhs) * value_of(rhs)

for v in result:
    print(v)
